core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java

```diff
@@ -87,6 +108,34 @@ public static BaseFileSystem create(FileSystemContext context) {
    */
   protected BaseFileSystem(FileSystemContext fsContext) {
     mFsContext = fsContext;
+    mBlockStore = AlluxioBlockStore.create(fsContext);
+  }
+
+  /**
+   * Shuts down the FileSystem. Closes all thread pools and resources used to perform operations. If
+   * any operations are called after closing the context the behavior is undefined.
+   *
+   * @throws IOException
+   */
+  @Override
+  public void close() throws IOException {
+    if (!mClosed) {
```

Comment:
nit: double-checking this feels like overkill when `close` is normally only called once

Reply:
`mClosed` needs to volatile? Though I agree we can just lock the whole method if close is not expected to be called repeatedly.

Reply:
It shouldn't be called repeatedly but I wanted to reduce waiting in the case it ever does get called. It was just for safety. Either way synchronizing the method would simplify the logic.